lib/spack/spack/package_base.py

```diff
@@ -373,6 +373,16 @@ def _flush_callbacks(check_name):
                             phase = getattr(base, phase_attr, None)
                             if phase is not None:
                                 break
+                        else:
+                            tty.die(
+                                """\
+bad directive @{directive}() for functions {funcs} in package {pkg}:""".format(
+                                    directive=check_name, funcs=funcs, pkg=name
+                                ),
+                                'no phase "{phase}" in any base {bases}'.format(
+                                    phase=phase_name, bases=bases
+                                ),
```

Comment:
I think we should avoid showing too much python internals as part of the error message. Something like this might be easier to interpret.
```suggestion
                            tty.die(
                                """\
bad directive @{directive}() for functions {funcs} in package {pkg}:""".format(
                                    directive=check_name, funcs=[f.__name__ for f in funcs], pkg=name
                                ),
                                'no phase "{phase}" in any {cls} or any base class thereof'.format(
                                    phase=phase_name, cls=cls
                                ),
```

Reply:
`cls` seems a little too low-level still: `<class 'spack.package_base.PackageMeta'>`: how about just `name` (`PyBadPackage`)?